snuba_sdk/query_visitors.py

```diff
@@ -0,0 +1,321 @@
+from __future__ import annotations
+
+import json
+from abc import ABC, abstractmethod
+from typing import (
+    Generic,
+    List,
+    MutableMapping,
+    Optional,
+    TypeVar,
+    TYPE_CHECKING,
+    Union,
+)
+
+from snuba_sdk.entity import Entity
+from snuba_sdk.conditions import Condition
+from snuba_sdk.expressions import (
+    Column,
+    Function,
+    Granularity,
+    Limit,
+    LimitBy,
+    Offset,
+    OrderBy,
+    Totals,
+)
+from snuba_sdk.visitors import Translation
+
+if TYPE_CHECKING:
+    from snuba_sdk.query import Query
+
+
+class InvalidQuery(Exception):
+    pass
+
+
+QVisited = TypeVar("QVisited")
+
+
+class QueryVisitor(ABC, Generic[QVisited]):
+    def visit(self, query: Query) -> QVisited:
+        returns = {
+            "dataset": self._visit_dataset(query.dataset),
+            "match": self._visit_match(query.match),
+            "select": self._visit_select(query.select),
+            "groupby": self._visit_groupby(query.groupby),
+            "where": self._visit_where(query.where),
+            "having": self._visit_having(query.having),
+            "orderby": self._visit_orderby(query.orderby),
+            "limitby": self._visit_limitby(query.limitby),
+            "limit": self._visit_limit(query.limit),
+            "offset": self._visit_offset(query.offset),
+            "granularity": self._visit_granularity(query.granularity),
+            "totals": self._visit_totals(query.totals),
+        }
```

Comment:
The clauses are statically defined, they will always be the same and always all visited. It seems this structure should be a NamedTuple instead of a dict.

Reply:
Changing this to a NamedTuple requires creating a Generic named tuple class and then creating a new version for every return type. I think this is simpler.

Reply:
A generic class can be concrete. Your named tuple does not need to be abstract:

In theory you should be able to do this.
```
TField = TypeVar("TField")
class PreCombineQuery(Generic[TField], NamedTuple): # Not sure which order the generic and named tuple should be in the declaration. And you can do the same with a dataclass.
   dataset: TField
   match: TField
   ..... 
```

In practice both named tuples and dataclasses need some more help to get the type right:
```
class PreCombineQuery(Generic[TField], NamedTuple):
       dataset: TField

       get_dataset(self) -> TField
```

Reply:
So I did try this, and while I could create the Generic class, I couldn't go further. Here's one example below with the mypy errors, although I tried a number of different ways to see if I could fix the errors.

```
from typing import TypeVar, Generic, NamedTuple


TField = TypeVar("TField")


class Parts(NamedTuple, Generic[TField]):
    part1: TField
    part2: TField

# "Parts" expects no type arguments, but 1 given
# Generic tuple types not supported
# Class cannot subclass value of type "Any"
class StrParts(Parts[str]):
    pass


p = StrParts("1", "2")
```

Reply:
@fpacifici I know we talked about a NamedTuple for this, but now that it's driven by the query itself I don't think we need an intermediate dataclass.

Reply:
ok